### PR TITLE
Fix context cancelled with reading git status operation

### PR DESCRIPTION
**Pull request by teresaromero**
# Proposed changes

Fixes: https://github.com/okteto/app/issues/6687

Relates: https://github.com/okteto/okteto/issues/3699

- Add custom Cancel function to `exec.CommandContext` so when the context is cancelled by external functions, the funcion exits with a termination signal instead of kill, that way git can remove the lock file and recover from the cancelation.

EDIT: after research, for windows we have to keep killing the process as there is no termination signal implementation

### CODE REVIEW

File: cmd/deploy/deploy.go
Comment (ifbyol) [inline cmd/deploy/deploy.go:99]: Define properties in a struct which won't be initialized as part of the instance creation is misleading. It might lead to panics (nil pointers. In this case it doesn't matter as it's not a pointer) or a wrong logic based on checks when the property wasn't set yet. I would try to use this pattern to pass the sha to the command configuration

File: cmd/deploy/deploy.go
Comment (ifbyol) [inline cmd/deploy/deploy.go:303]: Why are these changes required to fix the issue?

We should review where this is being called. Checking the code, it seems that during a deploy command execution, this function or the analogous `IsClean` might be called at least 3 times. Also, before this change, this logic was on `command_configuration` and it was being executed only when `OktetoGitCommitEnvVar` wasn't already set, so now, we will be calling it more times potentially impacting the performance.

If these changes are required to fix the issue (I don't see `repositorySHA` being used in the function which will kill the process), let's revert it.

File: pkg/repository/local_git.go
Comment (ifbyol) [inline pkg/repository/local_git.go:32]: I think CircleCI is complaining about the existence of `Cancel`

Comment (teresaromero) [reply pkg/repository/local_git.go:32]: yes, i am investigating

**Review by andreafalzetti — APPROVED** [review]

**Comment by teresaromero** [comment]
working parallel in https://github.com/okteto/okteto/tree/tere/isclean-sha-optimization_win to resolve the windows integration failing test when updating go

**Comment by maroshii** [comment]
Looking at the documentation and exec code, it looks like `cmd.Cancel` simply sends a signal to the parent process upon context cancellation. 

I haven't tested this but in case it helps we could do this ~manually (at least for testing) with something like this:

```golang
func (*LocalExec) RunCommand(ctx context.Context, dir string, name string, arg ...string) ([]byte, error) {
	c := exec.Command(name, arg...)
	c.Dir = dir
	c.Env = os.Environ()

	go func() {
		<-ctx.Done() // TODO: Add timeout so that this doesn't block forever if ctx is never cancelled
		oktetoLog.Debugf("terminating %s - %s/%s", c.String(), dir, name)
		if err := c.Process.Signal(syscall.SIGTERM); err != nil {
			oktetoLog.Debugf("err at signal SIGTERM: %v", err)
		}

		time.Sleep(5 * time.Second)
		if err := c.Process.Signal(syscall.Signal(0)); err != nil {
			if errors.Is(err, os.ErrProcessDone) {
				return nil
			}
			oktetoLog.Debugf("reading signal with error %v", err)
		}
		oktetoLog.Debugf("killing %s - %s/%s", c.String(), dir, name)
		return c.Process.Signal(syscall.SIGKILL)

	}()

	return c.Output()
}
```

**Review by AdrianPedriza — APPROVED** [review]

File: pkg/repository/local_git.go
Comment (jLopezbarb) [inline pkg/repository/local_git.go:34]: Do we know if this is working on windows? We have been facing issues with signals on windows before and leaving zombie processes because of that

File: pkg/repository/local_git.go
Comment (jLopezbarb) [inline pkg/repository/local_git.go:38]: Why 5 seconds? I'd try to avoid delaying it for that long

File: pkg/repository/local_git.go
Comment (jLopezbarb) [inline pkg/repository/local_git.go:32]: Could we somehow add a unit/integration test testing that this part of the code gets the proper execution?

Maybe a `RunCommand` of sleeping infinite and cancelling the context 1 ms after it's executing?

**Review by jLopezbarb — COMMENTED** [review]
Code looks good but check the comments before merging

**Comment by andreafalzetti** [comment]
@teresaromero I am moving this to draft until the other branch is ready, wdyt?

**Review by andreafalzetti — APPROVED** [review]

**Pull request merged**
